Allow changing the detail level of GCC dumps in TREE and RTL view

ibuclaw opened the pull request:
1. This Implements all options that are common for all versions of GCC between 4.0 and 10.2.
These are available in a "Options" drop-down menu on the GCC Tree Viewer pane.
   - address
   - slim
   - raw
   - details
   - stats
   - blocks
   - vops
   - lineno
   - uid
   - all

2. Dump passes have been moved into a drop-down menu "Passes" to keep the row tidy.

3. Add inter-procedural analysis dumps to GCC Tree Viewer.  The option -fdump-ipa-all has been around since GCC 4.0, though leaving it default unchecked in the UI.

4. Adds [ON] or [OFF] to the titles of all filter buttons by implementing GccDump.prototype.updateButtons.

Icons have been half randomly chosen.  For the Passes drop-down have used [fa-compass](https://fontawesome.com/v4.7.0/icon/compass), and for Options [fa-microchip](https://fontawesome.com/v4.7.0/icon/microchip).  If anyone has any better suggestions, I'm all ears.

Similarly, this lacks any tests, so if there's any suggestions on how best to add then, will have a look if it might be possible to add later to this pull.

Fixes #2130

ibuclaw commented:
> Similarly, this lacks any tests, so if there's any suggestions on how best to add then, will have a look if it might be possible to add later to this pull.

Tests are failing anyway, so will need to address that first.

apmorton commented on static/panes/compiler.js:
```diff
@@ -915,6 +916,28 @@ Compiler.prototype.onGccDumpFiltersChanged = function (id, filters, reqCompile)
         this.treeDumpEnabled = (filters.treeDump !== false);
         this.rtlDumpEnabled = (filters.rtlDump !== false);
 
+        this.dumpFlags = '';
```
not sure this makes sense to me - am I missing something?

If you check multiple options its going to produce a string like:
`-address-slim-raw`

ibuclaw replied:
It is correct usage.  From the docs

> If the ‘-options’ form is used, options is a list of ‘-’ separated options which control the details of the dump. Not all options are applicable to all dumps; those that are not meaningful are ignored.

apmorton replied:
ah, ok - I was looking at the docs for `-fdump-rtl`, which doesn't have the same note about combining in that way.

ibuclaw replied:
> ah, ok - I was looking at the docs for `-fdump-rtl`, which doesn't have the same note about combining in that way.

I can check it to be sure, but as I understand it, they all use the same machinery for option parsing.

ibuclaw replied:
Confirmed.  It's a bit of a free-for-all option.  The only part that is fixed is `-fdump-`, everything else that follows is handled by magic.

mattgodbolt commented:
Great stuff! Thanks for this! You are using the same time-honoured approach to icon choice that I've used too =)

mattgodbolt commented:
@RabsRincon / @apmorton any ideas what the "inaccessible to integration" CI failure is?

apmorton commented on static/panes/compiler.js:
```diff
@@ -550,6 +550,7 @@ Compiler.prototype.compile = function (bypassCache, newTools) {
                 pass: this.gccDumpPassSelected,
                 treeDump: this.treeDumpEnabled,
                 rtlDump: this.rtlDumpEnabled,
+                dumpFlags: this.dumpFlags,
```
I would rather we simply pass the bools for each option here and let the server side handle the actual command line arguments

ibuclaw replied:
I would rather have an enum flag instead of lots of booleans.

apmorton commented:
the CI failure is likely no fault of your own - we have made some changes recently that are causing PRs from outside contributors to fail - as long as `make lint && make test` pass on your machine you can safely ignore the failures here for now.

ibuclaw commented:
> the CI failure is likely no fault of your own - we have made some changes recently that are causing PRs from outside contributors to fail - as long as `make lint && make test` pass on your machine you can safely ignore the failures here for now.

OK, I'll check that and will fix up anything that gets flagged.

apmorton commented on lib/base-compiler.js:
```diff
@@ -477,10 +477,10 @@ class BaseCompiler {
 
         if (gccDumpOptions.treeDump !== false) {
-            addOpts.push('-fdump-tree-all');
+            addOpts.push('-fdump-tree-all' + gccDumpOptions.dumpFlags);
```
given the other conversation below, surely we need to remove the `-all` bit from this, yes?

apmorton replied:
I'm just going to bow out on this - tried it locally and it appears to be working.  The GCC docs are _incredibly_ confusing on this subject...

ibuclaw replied:
> given the other conversation below, surely we need to remove the `-all` bit from this, yes?

The answer would be no. The format is `-fdump-what-pass-options`. So `-fdump-tree-all-all` would produce dumps for all passes with all options enabled.

ibuclaw commented:
- Moved option builder to lib/base-compiler.js, not too thrilled to have multiple big chunks of boolean vars being copied around.
- Alphabetically reordered options in the drop-down.
- All options button has been moved to the top.
- Added a divider between Slim/Raw Dump and the other options.
- Whilst I am here looking at this code, I've added IPA pass dumps  (`-fdump-ipa-all`) as well (default unchecked in the UI).

apmorton commented on static/panes/compiler.js:
```diff
@@ -914,6 +925,18 @@ Compiler.prototype.onGccDumpFiltersChanged = function (id, filters, reqCompile)
         this.rtlDumpEnabled = (filters.rtlDump !== false);
+        this.ipaDumpEnabled = (filters.ipaDump !== false);
+
+        this.addressOptionEnabled = (filters.addressOption !== false);
```
If you want to avoid copying big chunks of bools we could build a single object here.

```javascript
this.dumpFlags = {
    slim: filters.slimOption !=== false,
    raw: filers.rawOption !=== false,
    ...etc
};
```

and then you just have to deal with a single `dumpFlags` object everywhere else.

apmorton commented:
side note: all PRs are now squashed when merging, so feel free to add commits instead of force pushing to make it clear what's changed between revisions

ibuclaw commented:
> side note: all PRs are now squashed when merging, so feel free to add commits instead of force pushing to make it clear what's changed between revisions

So I'll lose all my pretty commit messages? :-)

apmorton commented:
> So I'll lose all my pretty commit messages? :-)

We send them to a nice farm upstate

partouf commented:
> > So I'll lose all my pretty commit messages? :-)
> 
> We send them to a nice farm upstate

LIES!

ibuclaw commented:
Review notes:
- Moved all boolean options into a dumpFlags object.
- Squashed all commits into one big one.

Added Features:
- Dump filter buttons now have `[ON]` or `[OFF]` prepended to the titles (same as in source editor).

mattgodbolt commented on lib/base-compiler.js:
```diff
@@ -476,11 +476,49 @@ class BaseCompiler {
+        // GCC accepts these options as a list of '-' separated names that may
+        // appear in any order.
+        var flags = '';
+        if (gccDumpOptions.dumpFlags.address !== false) {
```
Is there a reason for the curiously specific `!== false` here? Is it specifically trying to avoid the normal "truthiness" of js objects?

mattgodbolt replied:
Forgive me: I see the original code was phrased this way! I don't know why, but as you can see I've merged and will update the site tonight!

Thank you again so much!

ibuclaw replied:
> Forgive me: I see the original code was phrased this way! I don't know why, but as you can see I've merged and will update the site tonight!
> 
> Thank you again so much!

Yep, I just followed the existing style.
